### PR TITLE
Project Recovery: Handle algorithms that require a string list of workspace names

### LINKED ISSUE #23452: Gracefully handle groupworkspaces instead of deleting them
### Expected behavior
GroupWorkspaces should be present after a recovery
### Actual behavior
GroupWorkspaces are not present
### Steps to reproduce the behavior
Group two workspaces and then wait for project recovery to save, Run Segfault, load up mantid and attempt recovery.
### Platforms affected
All

### PR DESCRIPTION
**Pull request by Pasarus**
Currently High priority because of a bug present in master with regards to project recovery that popped up recently and will be present in the nightly build if someone were to try recovering grouped workspaces mantid would tell them it is broken however this would not be the truth.

**Background of work**
GroupWorkspaces was originally removed from being recovered and it WAS working fine however due to a change recently in the way it checks whether workspaces are back which I suspect is due to a internal Qt change but could also be a master change, I am working on fixing the problem instead of the symptom.

Currently project recovery stores all history of workspaces in the files, this includes Saving to files, so the idea is to stop it from calling these saved files, the best way to leave the other algorithms still functioning normally was to stop it from saving specific files.

**Description of work.**
Removed the previous way of handling GroupWorkspaces as it will no longer after required pull requests be required as it is handled properly now. I also stopped the saving of save based algorithms so now it will not re-write files to disk on recovery

**To test:**
- Group two workspaces
- Wait for project recovery to save
- Run [`Segfault(DryRun=False)`](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/Segfault.cpp)
- Load up mantid
- Attempt recovery
- The group should be there with it's respective workspaces

and this test aswell

 - Load another workspace (From a nexus file would be easier)
 - Run SaveNexus on this workspace
 - Wait for Project Recovery to save (using the debug level of logging in the logger)
- Run [`Segfault(DryRun=False)`](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/Segfault.cpp)
- Reload mantid
 - Click open script in scripting window it should only have the Load algorithm calls, and not the  SaveNexus.

Fixes #23452 and #23450

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by keeeto** [comment]
The functional tests related to this PR work fine.

**Review by keeeto — APPROVED** [review]
This passes the functional tests and also the unscripted testing for project recovery.

File: Framework/API/inc/MantidAPI/ScriptBuilder.h [hunk 1]
```diff
@@ -49,7 +49,7 @@ class MANTID_API_DLL ScriptBuilder {
   ScriptBuilder(boost::shared_ptr<HistoryView> view,
                 std::string versionSpecificity = "old",
                 bool appendTimestamp = false,
-                bool ignoreGroupWorkspaces = false);
+                std::vector<std::string> ignoreTheseAlgs = {""});
```
Comment (soininen) [inline Framework/API/inc/MantidAPI/ScriptBuilder.h]: Why add an empty string? Why not use an empty `vector` as the default?

File: Framework/API/src/Algorithm.cpp [hunk 2]
```diff
@@ -1093,9 +1095,76 @@ void Algorithm::findWorkspaceProperties(
...
+      const auto propPropValue = propProp->value();
+      const auto direction = propProp->direction();
+      std::string currentWS = "";
+      for (auto i = 0u; i < propPropValue.size(); ++i) {
+        if (propPropValue[i] == ',') {
+          constructWorkspaceVectorForHistoryHelper(
+              inputWorkspaces, outputWorkspaces, direction, currentWS);
+          currentWS = "";
+        } else {
+          currentWS.push_back(propPropValue[i]);
```
Comment (soininen) [inline Framework/API/src/Algorithm.cpp]: `emplace_back` should be used over `push_back`.

File: Framework/API/src/Algorithm.cpp [hunk 3]
```diff
@@ -1093,9 +1095,76 @@ void Algorithm::findWorkspaceProperties(
...
+    std::vector<Workspace_sptr> &inputWorkspaces,
+    std::vector<Workspace_sptr> &outputWorkspaces, const unsigned int direction,
+    std::string &currentWS) const {
+  const auto &ADS = AnalysisDataService::Instance();
+  try {
+    if (direction == Direction::Input || direction == Direction::InOut) {
+      inputWorkspaces.push_back(ADS.retrieveWS<Workspace>(currentWS));
+    }
+  } catch (const Mantid::Kernel::Exception::NotFoundError &error) {
+    std::string errorMsg(error.what());
```
Comment (soininen) [inline Framework/API/src/Algorithm.cpp]: `const`

File: Framework/API/src/Algorithm.cpp [hunk 4]
```diff
@@ -1093,9 +1095,76 @@ void Algorithm::findWorkspaceProperties(
...
+    g_log.information("The ADS was unable to find the input workspaces "
+                      "when attaching history: " +
+                      errorMsg);
+  }
+  try {
+    if (direction == Direction::Output || direction == Direction::InOut) {
+      outputWorkspaces.push_back(ADS.retrieveWS<Workspace>(currentWS));
```
Comment (soininen) [inline Framework/API/src/Algorithm.cpp]: `emplace_back`

File: Framework/API/src/Algorithm.cpp [hunk 5]
```diff
@@ -1093,9 +1095,76 @@ void Algorithm::findWorkspaceProperties(
...
+    }
+  } catch (const Mantid::Kernel::Exception::NotFoundError &error) {
+    std::string errorMsg(error.what());
```
Comment (soininen) [inline Framework/API/src/Algorithm.cpp]: `const`

File: Framework/Algorithms/src/GeneratePythonScript.cpp [hunk 6]
```diff
@@ -65,8 +65,8 @@ void GeneratePythonScript::init() {
       boost::make_shared<StringListValidator>(saveVersions),
       "When to specify which algorithm version was used by Mantid.");
 
-  declareProperty("IgnoreGroups", false,
-                  "Whether or not groups need to be filtered for",
+  declareProperty("IgnoreTheseAlgs", std::vector<std::string>({""}),
```
Comment (soininen) [inline Framework/Algorithms/src/GeneratePythonScript.cpp]: Why `std::vector<std::string>()` would not work as the default value?

**Review by soininen — CHANGES_REQUESTED** [review]

File: Framework/API/src/Algorithm.cpp (hunk 2 above)
Comment (soininen) [reply Framework/API/src/Algorithm.cpp]: Okay, sorry for the noise, then!

**Review by soininen — APPROVED** [review]

**Pull request merged**